File: src/de/ax/powermode/power/management/ElementOfPowerContainer.scala
/*
 * Copyright 2015 Baptiste Mesta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.ax.powermode.power.management

import java.awt.event.{ComponentEvent, ComponentListener}
import java.awt.{Graphics, Point, Rectangle}
import javax.swing._

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.event._
import com.intellij.openapi.editor.{Editor, ScrollingModel}
import de.ax.powermode._
import de.ax.powermode.power.ElementOfPower
import de.ax.powermode.power.element.{
  PowerBam,
  PowerFlame,
  PowerIndicator,
  PowerSpark
}

import scala.collection.JavaConversions._

object ElementOfPowerContainer {
  private val logger = Logger.getInstance(this.getClass)
}

/**
  * @author Baptiste Mesta
  */
class ElementOfPowerContainer(editor: Editor)
    extends JComponent
    with ComponentListener
    with Power {

  import ElementOfPowerContainer.logger

  val myParent = editor.getContentComponent
  myParent.add(this)
  this.setBounds(myParent.getBounds)
  setVisible(true)
  myParent.addComponentListener(this)

  val shakeComponents = Seq(editor.getComponent, editor.getContentComponent)
  var elementsOfPower = Seq.empty[(ElementOfPower, (Int, Int))]
  var lastShake = System.currentTimeMillis()
  var shakeData = Option.empty[(Int, Int, Int, Int)]

  var lastUpdate = System.currentTimeMillis()

  var lastPositions = Seq.empty[(Point, Point)]
  editor.getCaretModel.addCaretListener(new CaretAdapter {

    def changeCarets: Unit = {
      lastPositions = {
        editor.getCaretModel.getAllCarets.map(caret => {
          (Util.getPoint(editor.offsetToVisualPosition(caret.getSelectionStart),
                         caret.getEditor),
           Util.getPoint(editor.offsetToVisualPosition(caret.getSelectionEnd),
                         caret.getEditor))
        })
      }
    }

    override def caretPositionChanged(e: CaretEvent): Unit = changeCarets

    override def caretAdded(e: CaretEvent): Unit = changeCarets

    override def caretRemoved(e: CaretEvent): Unit = changeCarets
  })

  editor.getDocument.addDocumentListener(new DocumentAdapter {
    override def documentChanged(e: DocumentEvent): Unit = {
      if (powerMode.isBamEnabled) {
        val shouldAnimate = {
          e.getNewFragment.length() > 100 ||
          e.getOldFragment.length() > 100 ||
          e.getOldFragment.toString.count('\n' == _) > 1 ||
          e.getNewFragment.toString.count('\n' == _) > 1
        }
        if (shouldAnimate) {
          val width = {
            val l = (e.getOldFragment.toString + e.getNewFragment.toString)
              .split("\n")
            if (l.nonEmpty) {
              (l.maxBy(_.length).length / 2.0) * editor.getLineHeight
            } else {
              0
            }
          }
          if (lastPositions.nonEmpty) {
            val myLastPositions = lastPositions
            SwingUtilities.invokeLater(() => {
              myLastPositions.foreach {
                case (a, b) =>
                  initializeAnimation(a, new Point(b.x, b.y), width)
              }
            })
          }
        }
      }
    }
  })

  def getAllCaretPositions: Seq[Point] = {
    editor.getCaretModel.getAllCarets.map(caret =>
      Util.getPoint(caret.getVisualPosition, caret.getEditor))
  }

  def updateElementsOfPower() {
    var delta = System.currentTimeMillis() - lastUpdate
    if (delta > (1000.0 / powerMode.frameRate) * 2)
      delta = 16
    lastUpdate = System.currentTimeMillis()
    val db: Double = 1000.0 / 16
    if (elementsOfPower.nonEmpty) {
      elementsOfPower =
        elementsOfPower.seq.filterNot(p => p._1.update((delta / db).toFloat))
      repaint()
    }
  }

  def addPowerIndicator(): Unit = {
    val indicatorWidth = 100
    elementsOfPower :+= (PowerIndicator(
      getMyBounds.width - 20 - indicatorWidth,
      getMyBounds.height - 20 - indicatorWidth,
      indicatorWidth,
      indicatorWidth,
      1000,
      editor), getScrollPosition)
  }

  def initializeAnimation(point: Point) {

    this.setBounds(getMyBounds)

    if (powerMode.isSparksEnabled) {
      addSparks(point)
    }
    if (powerMode.isFlamesEnabled) {
      addFlames(point)
    }

    if (powerMode.isShakeEnabled) {
      doShake(shakeComponents)
    }
    repaint()

  }

  def initializeAnimation(a: Point, b: Point, lineWidth: Double): Unit = {
    val x = a.x
    var y = a.y
    val width = math.max(b.x - x, 50)
    val height = math.max(b.y - y, 50)
    val dim =
      (Seq(lineWidth.toInt, width, height, 50).max * powerMode.valueFactor).toInt
    if (b.y - y.abs < dim) {
      y = y - dim / 2
    }
    elementsOfPower :+= (PowerBam(
      x,
      y,
      dim,
      dim,
      powerMode.bamLife * powerMode.valueFactor toLong), getScrollPosition)
  }

  def addFlames(point: Point): Unit = {
    val base = 0.3
    val wh = (powerMode.maxFlameSize * base +
      ((math.random * powerMode.maxFlameSize * (1 - base)) * powerMode.valueFactor)).toInt
    val initLife = (powerMode.maxFlameLife * powerMode.valueFactor).toInt
    if (initLife > 100) {
      elementsOfPower :+= (PowerFlame(point.x + 5,
                                      point.y - 1,
                                      wh,
                                      wh,
                                      initLife,
                                      true), getScrollPosition)
      elementsOfPower :+= (PowerFlame(point.x + 5,
                                      point.y + 15,
                                      wh,
                                      wh,
                                      initLife,
                                      false), getScrollPosition)
    }
  }

  def addSparks(point: Point): Unit = {
    for (i <- 0 to (powerMode.sparkCount * powerMode.valueFactor).toInt) {
      addSpark(point.x, point.y)
    }
  }

  def addSpark(x: Int, y: Int) {
    val dx
      : Double = (Math.random * 2) * (if (Math.random > 0.5) -1 else 1) * powerMode.sparkVelocityFactor
    val dy: Double = ((Math.random * -3) - 1) * powerMode.sparkVelocityFactor
    val size = ((Math.random * powerMode.sparkSize) + 1).toInt
    val life = Math.random() * powerMode.getSparkLife * powerMode.valueFactor
    val powerSpark = PowerSpark(x,
                                y,
                                dx.toFloat,
                                dy.toFloat,
                                size,
                                life.toLong,
                                genNextColor,
                                powerMode.gravityFactor.toFloat)
    elementsOfPower :+= (powerSpark, getScrollPosition)
  }

  def genNextColor: PowerColor =
    (getColorPart(powerMode.getRedFrom, powerMode.getRedTo),
     getColorPart(powerMode.getGreenFrom, powerMode.getGreenTo),
     getColorPart(powerMode.getBlueFrom, powerMode.getBlueTo),
     powerMode.getColorAlpha / 255f)

  def getColorPart(from: Int, to: Int): Float = {
    (((Math.random() * (to - from)) + from) / 255).toFloat
  }

  def getScrollPosition = (
    editor.getScrollingModel.getHorizontalScrollOffset,
    editor.getScrollingModel.getVerticalScrollOffset
  )

  def doShake(myShakeComponents: Seq[JComponent]): Unit = {
    if (Util.editorOk(editor, 100)) {
      shakeData = shakeData match {
        case Some((dx, dy, scrollX, scrollY)) =>
          myShakeComponents.foreach { myShakeComponent =>
            val bounds: Rectangle = myShakeComponent.getBounds
            myShakeComponent.setBounds(bounds.x + dx,
                                       bounds.y + dy,
                                       bounds.width,
                                       bounds.height)
          }
          None
        case None =>
          val dx = generateShakeOffset
          val dy = generateShakeOffset
          val scrollX = editor.getScrollingModel.getHorizontalScrollOffset
          val scrollY = editor.getScrollingModel.getVerticalScrollOffset
          myShakeComponents.foreach { myShakeComponent =>
            val bounds: Rectangle = myShakeComponent.getBounds
            myShakeComponent.setBounds(bounds.x + dx,
                                       bounds.y + dy,
                                       bounds.width,
                                       bounds.height)
          }
          Some((-dx, -dy, scrollX, scrollY))
      }
      lastShake = System.currentTimeMillis()
    }
  }

  def generateShakeOffset: Int = {
    val range = powerMode.shakeRange * powerMode.valueFactor
    (range - (Math.random * 2 * range)).toInt
  }

  def componentResized(e: ComponentEvent) {
    setBounds(getMyBounds)
    logger.debug("Resized")
  }

  def componentMoved(e: ComponentEvent) {
    setBounds(getMyBounds)
    logger.debug("Moved")
  }

  def getMyBounds: Rectangle = {
    val area = editor.getScrollingModel.getVisibleArea
    val rectangle = new Rectangle(area.x, area.y, area.width, area.height)
    rectangle
  }

  def componentShown(e: ComponentEvent) {}

  def componentHidden(e: ComponentEvent) {}

  protected override def paintComponent(g: Graphics) {
    super.paintComponent(g)
    if (powerMode.isEnabled) {
      if (shakeData.isDefined &&
          System.currentTimeMillis() - lastShake > 100 &&
          shakeData.get._1.abs < 50 && shakeData.get._2.abs < 50) {
        doShake(Seq(editor.getComponent))
      }
      renderElementsOfPower(g)
    }
  }

  def renderElementsOfPower(g: Graphics) {
    val scrollingModel: ScrollingModel = editor.getScrollingModel
    val xyNew = (
      scrollingModel.getHorizontalScrollOffset,
      scrollingModel.getVerticalScrollOffset
    )

    elementsOfPower.foreach { pp =>
      val (elementOfPower, (x, y)) = pp
      val dxx = x - xyNew._1
      val dyy = y - xyNew._2
      elementOfPower.render(g, dxx, dyy)
    }

  }
}

File: src/com/nmeylan/powermode/power/management/ElementOfPowerContainer.java
package com.nmeylan.powermode.power.management;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.ScrollingModel;
import com.intellij.openapi.editor.event.CaretAdapter;
import com.intellij.openapi.editor.event.CaretEvent;
import com.intellij.openapi.editor.event.DocumentAdapter;
import com.intellij.openapi.editor.event.DocumentEvent;
import com.nmeylan.powermode.Pair;
import com.nmeylan.powermode.Power;
import com.nmeylan.powermode.Util;
import com.nmeylan.powermode.power.ElementOfPower;
import com.nmeylan.powermode.power.element.PowerBam;
import com.nmeylan.powermode.power.element.PowerFlame;
import com.nmeylan.powermode.power.element.PowerSpark;
import org.jetbrains.annotations.NotNull;

import javax.swing.JComponent;
import javax.swing.SwingUtilities;
import java.awt.Graphics;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.event.ComponentEvent;
import java.awt.event.ComponentListener;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;


public class ElementOfPowerContainer extends JComponent implements ComponentListener, Power {

  private Editor editor;

  private List<JComponent> shakeComponents;
  private List<Pair<ElementOfPower, Point>> elementsOfPower;
  private long lastShake;
  private long lastUpdate;
  private List<Point> shakeData;
  private List<List<Point>> lastPositionsPerCarets;

  public ElementOfPowerContainer(Editor editor) {
    super();
    this.editor = editor;
    this.shakeComponents = Arrays.asList(editor.getComponent(), editor.getContentComponent());
    this.elementsOfPower = new ArrayList<>();
    this.lastPositionsPerCarets = new ArrayList<>();
    this.shakeData = new ArrayList<>();
    this.lastShake = System.currentTimeMillis();
    this.lastUpdate = System.currentTimeMillis();
    JComponent myParent = editor.getContentComponent();
    myParent.add(this);
    this.setBounds(myParent.getBounds());
    setVisible(true);
    myParent.addComponentListener(this);
    editor.getCaretModel().addCaretListener(new CaretAdapter() {
      public void changeCarets() {
        lastPositionsPerCarets = editor.getCaretModel().getAllCarets().stream().map(
          caret -> Arrays.asList(
            Util.getPoint(editor.offsetToVisualPosition(caret.getSelectionStart()), caret.getEditor()),
            Util.getPoint(editor.offsetToVisualPosition(caret.getSelectionEnd()), caret.getEditor()))).collect(Collectors.toList());
      }

      @Override
      public void caretPositionChanged(@NotNull CaretEvent event) {
        changeCarets();
      }

      @Override
      public void caretAdded(@NotNull CaretEvent event) {
        changeCarets();
      }

      @Override
      public void caretRemoved(@NotNull CaretEvent event) {
        changeCarets();
      }
    });

    editor.getDocument().addDocumentListener(new DocumentAdapter() {
      @Override
      public void documentChanged(@NotNull DocumentEvent event) {
        if (powerMode().isBamEnabled()) {
          boolean shouldAnimate = (
            event.getNewFragment().length() > 100 ||
              event.getOldFragment().length() > 100 ||
              event.getOldFragment().toString().contains("\n") ||
              event.getNewFragment().toString().contains("\n")
          );
          if (shouldAnimate) {
            List<String> lineFeeds = Arrays.asList((event.getOldFragment().toString() + event.getNewFragment().toString()).split("\n"));
            int width = lineFeeds.isEmpty() ? 0 : (int) (lineFeeds.stream().max(Comparator.comparing(String::length)).get().length() / 2.0) * editor.getLineHeight();
            if (!lastPositionsPerCarets.isEmpty()) {
              SwingUtilities.invokeLater(() -> lastPositionsPerCarets.stream()
                .forEach(caretPositions -> initializeAnimation(caretPositions.get(0), caretPositions.get(1), width)));
            }
          }
        }
      }

    });
  }


  private List<Point> getAllCaretPositions() {
    return editor.getCaretModel().getAllCarets().stream()
      .map(caret -> Util.getPoint(caret.getVisualPosition(), caret.getEditor()))
      .collect(Collectors.toList());
  }

  public void updateElementsOfPower() {
    long delta = System.currentTimeMillis() - lastUpdate;
    if (delta > (1000.0 / powerMode().getFrameRate()) * 2) {
      delta = 16;
    }
    lastUpdate = System.currentTimeMillis();
    double db = 1000.0 / 16;
    long deltaa = delta;
    if (!elementsOfPower.isEmpty()) {
      elementsOfPower =
        elementsOfPower.stream().filter(p -> {
          long start = System.nanoTime();
          p.first().update((deltaa / db));

          long diff = (System.nanoTime() - start);
          System.out.println(diff / 1000.0 + " ms");
          return p.first().alive();
        }).collect(Collectors.toList());
      repaint();
    }
  }

  public void initializeAnimation(Point point) {
    this.setBounds(getMyBounds());

    if (powerMode().isSparksEnabled()) {
      addSparks(point);
    }
    if (powerMode().isFlamesEnabled()) {
      addFlames(point);
    }

    if (powerMode().isShakeEnabled()) {
      doShake(shakeComponents);
    }
    repaint();
  }

  private void initializeAnimation(Point a, Point b, int lineWidth) {
    int x = a.x;
    int y = a.y;
    int width = Math.max(b.x - x, 50);
    int height = Math.max(b.y - y, 50);
    int dim = (int) (Arrays.asList(lineWidth, width, height, 50).stream().mapToInt(i -> i).max().getAsInt() * powerMode().valueFactor());
    if (b.y - Math.abs(y) < dim) {
      y = y - dim / 2;
    }
    elementsOfPower.add(
      Pair.with(
        new PowerBam(x, y, dim, dim, (long) (powerMode().getBamLife() * powerMode().valueFactor())),
        getScrollPosition()));
  }

  private void addFlames(Point point) {
    float base = 0.3f;
    int wh = (int) ((powerMode().getMaxFlameSize() * base +
      ((Math.random() * powerMode().getMaxFlameSize() * (1 - base)) * powerMode().valueFactor())));
    int initLife = (int) (powerMode().getMaxFlameLife() * powerMode().valueFactor());
    if (initLife > 100) {
      elementsOfPower.add(
        Pair.with(
          new PowerFlame(point.x + 5, point.y - 1, wh, wh, initLife, true),
          getScrollPosition()));
      elementsOfPower.add(
        Pair.with(
          new PowerFlame(point.x + 5, point.y + 15, wh, wh, initLife, false),
          getScrollPosition()));
    }
  }

  private void addSparks(Point point) {
    for (int i = 0; i < (int) (powerMode().getSparkCount() * powerMode().valueFactor()); i++) {
      addSpark(point.x, point.y);
    }
  }

  private void addSpark(int x, int y) {
    float dx = (float) ((Math.random() * 2) * (Math.random() > 0.5 ? -1 : 1) * powerMode().getSparkVelocityFactor());
    float dy = (float) (((Math.random() * -3) - 1) * powerMode().getSparkVelocityFactor());
    int size = (int) ((Math.random() * powerMode().getSparkSize()) + 1);
    int life = (int) (Math.random() * powerMode().getSparkLife() * powerMode().valueFactor());
    elementsOfPower.add(
      Pair.with(
        new PowerSpark(x,
          y,
          dx,
          dy,
          size,
          life,
          genNextColor(),
          (float) powerMode().getGravityFactor()),
        getScrollPosition()
      ));
  }

  private float[] genNextColor() {
    return new float[]{
      getColorPart(powerMode().getRedFrom(), powerMode().getRedTo()),
      getColorPart(powerMode().getGreenFrom(), powerMode().getGreenTo()),
      getColorPart(powerMode().getBlueFrom(), powerMode().getBlueTo()),
      powerMode().getColorAlpha() / 255f};
  }

  private float getColorPart(int from, int to) {
    return (float) (((Math.random() * (to - from)) + from) / 255);
  }

  private Point getScrollPosition() {
    return new Point(editor.getScrollingModel().getHorizontalScrollOffset(),
      editor.getScrollingModel().getVerticalScrollOffset());

  }

  private void doShake(List<JComponent> myShakeComponents) {
    if (Util.editorOk(editor, 100)) {
      int x, y;
      if (shakeData.size() >= 1) {
        x = shakeData.get(0).x;
        y = shakeData.get(0).y;
        shakeData.clear();
      } else {
        x = generateShakeOffset();
        y = generateShakeOffset();
        int scrollX = editor.getScrollingModel().getHorizontalScrollOffset();
        int scrollY = editor.getScrollingModel().getVerticalScrollOffset();
        shakeData.add(new Point(x, y));
        shakeData.add(new Point(scrollX, scrollY));
      }
      myShakeComponents.stream().forEach(component -> {
        Rectangle bounds = component.getBounds();
        component.setBounds(
          bounds.x + x,
          bounds.y + y,
          bounds.width,
          bounds.height
        );
      });
      lastShake = System.currentTimeMillis();
    }
  }

  private int generateShakeOffset() {
    int range = (int) (powerMode().getShakeRange() * powerMode().valueFactor());
    return (int) (range - (Math.random() * 2 * range));
  }

  public void componentResized(ComponentEvent e) {
    setBounds(getMyBounds());
    powerMode().logger().debug("Resized");
  }

  public void componentMoved(ComponentEvent e) {
    setBounds(getMyBounds());
    powerMode().logger().debug("Moved");
  }

  @Override
  public void componentShown(ComponentEvent e) {

  }

  @Override
  public void componentHidden(ComponentEvent e) {

  }

  private Rectangle getMyBounds() {
    Rectangle area = editor.getScrollingModel().getVisibleArea();
    return new Rectangle(area.x, area.y, area.width, area.height);
  }

  @Override
  protected void paintComponent(Graphics g) {
    super.paintComponent(g);
    if (powerMode().isEnabled()) {
      if (shakeData != null && shakeData.size() >= 2 &&
        System.currentTimeMillis() - lastShake > 100 &&
        Math.abs(shakeData.get(0).x) < 50 && Math.abs(shakeData.get(1).y) < 50) {
        doShake(Arrays.asList(editor.getComponent()));
      }
      renderElementsOfPower(g);
    }
  }

  private void renderElementsOfPower(Graphics g) {
    ScrollingModel scrollingModel = editor.getScrollingModel();
    Point newElementPosition = new Point(scrollingModel.getHorizontalScrollOffset(), scrollingModel.getVerticalScrollOffset());

    elementsOfPower.stream().forEach(elementOfPowerPointPair -> {
      int x = elementOfPowerPointPair.last().x - newElementPosition.x;
      int y = elementOfPowerPointPair.last().y - newElementPosition.y;
      elementOfPowerPointPair.first().render(g, x, y);
    });

  }
}